### PR TITLE
Refactor mnist hello world example

### LINKED ISSUE #240: Refactor Mnist Hello World Example
### Motivation

I have refactored the entire mnist hello world example, maintaining a similar structure but much more readable. 

### Pitch

The following are the major changes:
- Addition of a dataclass config instead of hardcoded config values
- Change in the implementation style of the forward function of MNISTModel 
- Add the `ckpt_path` argument to `trainer.test()` to suppress the warning
- Addition of docstrings to all classes and methods
- Addition of type hints for all parameters
- Addition of comments as and when required

### Additional context

I am a member of the Lightning League and the refactoring has been discussed with @Borda

### PR DESCRIPTION
**Pull request by ishandutta0098**
## What does this PR do?
I have refactored the entire mnist hello world example, maintaining a similar structure but much more readable.

Following are the major changes:

1. Addition of a `dataclass` config instead of hardcoded config values
2. Change in the implementation style of the `forward()` function of `MNISTModel`
3. Add the `ckpt_path` argument to `trainer.test()` to suppress the warning
4. Addition of docstrings to all classes and methods
5. Addition of type hints for all parameters
6. Addition of comments as and when required

## Additional Information:
- I have used patched commits to ensure that the code review becomes easier. 
- Code formatting has been done using black
- The code has been tested via a local run
  
## PR review

I am a member of the Lightning League and the refactoring has been discussed with @Borda

closes #240

### CODE REVIEW

File: lightning_examples/mnist-hello-world/hello-world.py [hunk 1]
```diff
@@ -97,8 +201,70 @@ def configure_optimizers(self):
...
 
 # %%
+
+
 class LitMNIST(L.LightningModule):
-    def __init__(self, data_dir=PATH_DATASETS, hidden_size=64, learning_rate=2e-4):
+    """PyTorch Lightning module for training a multi-layer perceptron (MLP) on the MNIST dataset.
+
+    Attributes
+    ----------
```
Comment (Borda) [inline lightning_examples/mnist-hello-world/hello-world.py:210]: lets unify the documentation style and mostly we use Napoleon

File: lightning_examples/mnist-hello-world/hello-world.py [hunk 2]
```diff
@@ -26,20 +71,77 @@
...
+        activated = torch.relu(hidden)
 
-    def training_step(self, batch, batch_nb):
+        return activated
+
+    def training_step(self, batch: Tuple[torch.Tensor, torch.Tensor], batch_nb: int) -> torch.Tensor:
+        """Defines a single training step for the model.
+
+        Args:
+            batch (Tuple[torch.Tensor, torch.Tensor]): A tuple containing the input and target tensors for the batch.
```
Comment (Borda) [inline lightning_examples/mnist-hello-world/hello-world.py:129]:
```suggestion
            batch: A tuple containing the input and target tensors for the batch.
```
Typing are already the method footprint so lets not duplicate the info

Comment (ishandutta0098) [reply lightning_examples/mnist-hello-world/hello-world.py:129]: Shall I remove the datatype in docstrings for all parameters whose type hints are present?

File: lightning_examples/mnist-hello-world/hello-world.py (hunk 1 above)
Comment (ishandutta0098) [reply lightning_examples/mnist-hello-world/hello-world.py]: I have changed the documentation style to Napoleon Google Format

File: lightning_examples/mnist-hello-world/hello-world.py (hunk 2 above)
Comment (ishandutta0098) [reply lightning_examples/mnist-hello-world/hello-world.py]: I have removed the typehints form the docstrings

**Review by Borda — APPROVED** [review]

**Pull request merged**